theano/tensor/nnet/opt.py

```diff
@@ -41,3 +55,299 @@ def local_inplace_sparse_block_outer(node):
                            local_inplace_sparse_block_outer,
                            failure_callback=gof.TopoOptimizer.warn_inplace),
                        60, 'fast_run', 'inplace')  # DEBUG
+
+
+# Conv opts
+@local_optimizer([AbstractConv2d])
+def local_abstractconv_gemm(node):
+    if not isinstance(node.op, AbstractConv2d):
+        return None
+    img, kern = node.inputs
+    if not isinstance(img.type, TensorType) or \
+       not isinstance(kern.type, TensorType):
+        return None
+
+    # need to flip the kernel if necessary
+    if node.op.filter_flip:
+        kern = kern[:, :, ::-1, ::-1]
+    rval = CorrMM(border_mode=node.op.border_mode,
+                  subsample=node.op.subsample)(img, kern)
+
+    return [rval]
+
+
+@local_optimizer([AbstractConv2d_gradWeights])
+def local_abstractconv_gradweight_gemm(node):
+    if not isinstance(node.op, AbstractConv2d_gradWeights):
+        return None
+    img, topgrad, shape = node.inputs
+    if not isinstance(img.type, TensorType) or \
+       not isinstance(topgrad.type, TensorType):
+        return None
+
+    rval = CorrMM_gradWeights(border_mode=node.op.border_mode,
+                              subsample=node.op.subsample)(img, topgrad, shape)
+    # need to flip the kernel if necessary
+    if node.op.filter_flip:
+        rval = rval[:, :, ::-1, ::-1]
+    rval = theano.tensor.patternbroadcast(rval, node.outputs[0].broadcastable)
+
+    return [rval]
+
+
+@local_optimizer([AbstractConv2d_gradInputs])
+def local_abstractconv_gradinputs_gemm(node):
+    if not isinstance(node.op, AbstractConv2d_gradInputs):
+        return None
+    kern, topgrad, shape = node.inputs
+    if not isinstance(kern.type, TensorType) or \
+       not isinstance(topgrad.type, TensorType):
+        return None
+
+    # need to flip the kernel if necessary
+    if node.op.filter_flip:
+        kern = kern[:, :, ::-1, ::-1]
+    rval = CorrMM_gradInputs(border_mode=node.op.border_mode,
+                             subsample=node.op.subsample)(kern, topgrad,
+                                                          shape)
+
+    return [rval]
+
+
+@local_optimizer([AbstractConv2d])
+def local_conv2d_cpu(node):
+
+    if not isinstance(node.op, AbstractConv2d):
+        return None
+
+    img, kern = node.inputs
+    if ((not isinstance(img.type, TensorType) or
+         not isinstance(kern.type, TensorType))):
+        return None
+    if node.op.border_mode not in ['full', 'valid']:
+        return None
+    if not node.op.filter_flip:
+        # Not tested yet
+        return None
+
+    rval = conv2d(img, kern,
+                  node.op.imshp, node.op.kshp,
+                  border_mode=node.op.border_mode,
+                  subsample=node.op.subsample)
+    return [rval]
+
+
+@local_optimizer([AbstractConv2d_gradWeights])
+def local_conv2d_gradweight_cpu(node):
+
```

Comment:
I think I understand where the issue comes from. Here, we should still check that `node` is a `..._gradWeights`, otherwise it could get applied to another node. When I ran the tests on the master branch, the optimizations were applied in the right order so that did not matter, but now that bug is uncovered.